File: kotlin/src/FizzBuzz.kt
/**
 * @author Artur Bosch
 */

fun main(args: Array<String>) {
	fizzbuzz(100)
}

fun fizzbuzz(n: Int): Unit {

	fun Int.dividesBy(n: Int): Boolean = this % n == 0

	fun fizzOrBuzz(it: Int): Any {
		return when {
			it.dividesBy(15) -> "FizzBuzz"
			it.dividesBy(5) -> "Buzz"
			it.dividesBy(3) -> "Fizz"
			else -> it
		}
	}

	(1..n).forEach {
		println(fizzOrBuzz(it))
	}
}

File: scala/src/FizzBuzz.scala
/**
  * @author Artur Bosch
  */
object FizzBuzz {

  implicit class NumberExt(number: Int) {
    def dividesBy(n: Int): Boolean = number % n == 0
  }

  def fizzbuzz(n: Int): Unit = {

    def cases(i: Int): String = {
      if (i dividesBy 15) return "fizzbuzz"
      if (i dividesBy 5) return "buzz"
      if (i dividesBy 3) "fizz" else i.toString
    }

    (1 to n).foreach(i =>
      println(cases(i))
    )
  }

  def main(args: Array[String]): Unit = {
    fizzbuzz(100)
  }
}
